File: java/programmers/src/문자열_압축.java
import java.util.*;
public class 문자열_압축 {
    class Solution {
        public int solution(String s) {
            int answer = s.length();
            for(int k=1;k<s.length();k++){
                int tmpAnswer = s.length();
                for(int index =0;index<s.length()-k;index+=k){
                    String tmpString = s.substring(index,index+k);
                    int cnt =1;
                    while(index+2*k<=s.length()&& s.substring(index+k,index+2*k).equals(tmpString))                 {
                        cnt++;
                        index +=k;
                    }
                    tmpAnswer -= ((cnt-1)*k);
                    if(cnt!=1){
                        tmpAnswer += Integer.toString(cnt).length();
                    }

                }
                answer = Math.min(answer,tmpAnswer);
            }
            return answer;
        }
    }
}

File: java/baekjoon/src/_문자열_압축.kt
import kotlin.math.min

class _문자열_압축 {
    class Solution {
        var answer = 0

        fun solution(s: String): Int {
            answer = s.length

            for (size in 1..s.length - 1) {
                var len = s.length
                var i = 0
                while (i <= s.length - size) {
                    val curS = s.substring(i until i + size)
                    var count = 1

//                    println("string is $curS ")
//                    println("end index of sToCompare is ${i + (size * 2) - 1}")

                    var j = i + size
                    while (true) {
                        i += size
                        if (j + size <= s.length) {
                            val sToCompare = s.substring(i..i + size - 1)
//                        println("string to Compare is $sToCompare")
                            if (sToCompare == curS) {
                                count++
                                j += size
//                            println("$s is equal to index ${i + size} to ${i + (size * 2) - 1} with $count times")
                            } else {
                                break
                            }
                        } else {
                            val tail = s.substring(j)
                            if (tail.isNotEmpty() && curS.startsWith(tail)) {
                                // 오해 압축 방지: 3aba 같은 케이스 → 압축 취소
                                count = 1
                            }
                            break
                        }
                    }

                    i = j

                    if (count > 1) {
                        len = len - (size * count) + (count.toString().length + size)
                    }
                }

                answer = min(len, answer)
            }

            return answer
        }
    }
}
